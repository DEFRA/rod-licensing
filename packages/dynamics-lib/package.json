{
  "name": "@defra-fish/dynamics-lib",
  "version": "1.22.0-rc.3",
  "description": "Framework to support integration with dynamics",
  "type": "module",
  "engines": {
    "node": ">=14"
  },
  "keywords": [
    "rod",
    "licensing",
    "dynamics"
  ],
  "author": "DEFRA",
  "license": "SEE LICENSE IN LICENSE",
  "main": "src/index.js",
  "directories": {
    "lib": "src",
    "test": "__tests__"
  },
  "files": [
    "src"
  ],
  "repository": {
    "type": "git",
    "url": "git://github.com/DEFRA/rod-licensing.git",
    "directory": "packages/dynamics-lib"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "dependencies": {
    "cache-manager": "^3.4.3",
    "cache-manager-ioredis": "^2.1.0",
    "debug": "^4.3.1",
    "dynamics-web-api": "1.7.3",
    "joi": "^17.4.0",
    "moment": "^2.29.1",
    "pluralize": "^8.0.0",
    "simple-oauth2": "^4.2.0",
    "uuid": "^8.3.2"
  }
}
