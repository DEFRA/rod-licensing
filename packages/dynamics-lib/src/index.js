// Entities
export * from './entities/contact.entity.js'
export * from './entities/permission.entity.js'
export * from './entities/permit.entity.js'
export * from './entities/permit-concession.entity.js'
export * from './entities/concession.entity.js'
export * from './entities/concession-proof.entity.js'
export * from './entities/fulfilment-request.entity.js'
export * from './entities/fulfilment-request-file.entity.js'
export * from './entities/transaction.entity.js'
export * from './entities/transaction-journal.entity.js'
export * from './entities/transaction-currency.entity.js'
export * from './entities/pocl-file.entity.js'
export * from './entities/pocl-staging-exception.entity.js'
export * from './entities/pocl-data-validation-error.entity.js'
export * from './entities/recurring-payment.entity.js'
export * from './entities/recurring-payment-instruction.entity.js'
export * from './entities/staging-exception.entity.js'

export * from './entities/role.entity.js'
export * from './entities/system-user-role.entity.js'
export * from './entities/system-user.entity.js'

// Queries
export * from './queries/permission.queries.js'
export * from './queries/fulfilment.queries.js'

// Framework functionality
export * from './client/util.js'
export { dynamicsClient } from './client/dynamics-client.js'
export * from './client/cache.js'
export * from './client/entity-manager.js'
