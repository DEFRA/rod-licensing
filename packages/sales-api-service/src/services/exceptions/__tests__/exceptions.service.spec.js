import { createStagingException, createStagingExceptionFromError, createTransactionFileException } from '../exceptions.service.js'
import { persist, PoclStagingException, StagingException } from '@defra-fish/dynamics-lib'

jest.mock('@defra-fish/dynamics-lib', () => ({
  ...jest.requireActual('@defra-fish/dynamics-lib'),
  persist: jest.fn()
}))

expect.extend({
  jsonMatching (received, ...matchers) {
    try {
      const obj = JSON.parse(received)
      for (const matcher of matchers) {
        if (!matcher.asymmetricMatch(obj)) {
          return {
            message: () => `expected ${matcher.toString()} to pass`,
            pass: false
          }
        }
      }
      return {
        pass: true
      }
    } catch (e) {
      return {
        message: () => 'expected a valid json structure',
        pass: false
      }
    }
  }
})

describe('exceptions service', () => {
  describe('createStagingException', () => {
    it('creates a staging exception entity in Dynamics', async () => {
      const stagingException = {
        stagingId: 'string',
        description: 'string',
        transactionJson: 'string',
        exceptionJson: 'string'
      }
      const result = await createStagingException(stagingException)
      expect(persist).toHaveBeenCalledWith([expect.objectContaining(stagingException)])
      expect(result).toBeInstanceOf(StagingException)
    })
  })

  describe('createStagingExceptionFromError', () => {
    it('creates a staging exception entity in Dynamics by parsing the error message', async () => {
      const expectedErrorMessage = 'Error: Fake test error'
      const testError = new Error('Fake test error')
      const testTransaction = { some: 'data' }
      const result = await createStagingExceptionFromError('testStagingId', testError, testTransaction)
      expect(persist).toHaveBeenCalledWith([
        expect.objectContaining({
          stagingId: 'testStagingId',
          description: expectedErrorMessage,
          transactionJson: expect.jsonMatching(expect.objectContaining(testTransaction)),
          exceptionJson: expect.jsonMatching(
            expect.objectContaining({
              stack: expect.arrayContaining([expectedErrorMessage, ...testError.stack.split('\n')])
            })
          )
        })
      ])
      expect(result).toBeInstanceOf(StagingException)
    })

    it('applies different error parsing rules to exceptions originating from a Dynamics call', async () => {
      const expectedErrorMessage = 'Custom dynamics error message'
      const testError = Object.assign(new Error(), { error: { message: expectedErrorMessage } })
      const testTransaction = { some: 'data' }
      const result = await createStagingExceptionFromError('testStagingId', testError, testTransaction)
      expect(persist).toHaveBeenCalledWith([
        expect.objectContaining({
          stagingId: 'testStagingId',
          description: expectedErrorMessage,
          transactionJson: expect.jsonMatching(expect.objectContaining(testTransaction)),
          exceptionJson: expect.jsonMatching(
            expect.objectContaining({
              error: expect.objectContaining({ message: expectedErrorMessage }),
              stack: expect.arrayContaining([...testError.stack.split('\n')])
            })
          )
        })
      ])
      expect(result).toBeInstanceOf(StagingException)
    })
  })

  describe('createTransactionFileException', () => {
    it('creates a pocl staging exception entity in Dynamics', async () => {
      const transactionFileException = {
        name: 'string',
        description: 'string',
        json: 'string',
        notes: 'string',
        type: 'Failure',
        transactionFile: 'string',
        permissionId: 'string'
      }
      const result = await createTransactionFileException(transactionFileException)
      expect(persist).toHaveBeenCalledWith([
        expect.objectContaining({
          ...transactionFileException,
          type: expect.objectContaining({ id: 910400001, label: 'Failure', description: 'Failure' }),
          status: expect.objectContaining({ id: 910400000, label: 'Open', description: 'Open' })
        })
      ])
      expect(result).toBeInstanceOf(PoclStagingException)
    })
  })
})
