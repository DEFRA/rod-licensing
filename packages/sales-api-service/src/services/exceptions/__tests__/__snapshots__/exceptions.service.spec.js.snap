// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exceptions service createDataValidationError maps the record data correctly 1`] = `
Object {
  "amount": 30,
  "birthDate": "1989-07-01",
  "channelId": "948594",
  "concessions": "[{\\"type\\":\\"Blue Badge\\",\\"referenceNumber\\":\\"123456789\\"}]",
  "country": "GB",
  "dataSource": Object {
    "description": "Post Office Sales",
    "id": 910400000,
    "label": "Post Office Sales",
  },
  "email": "daniel-ricc@example.com",
  "firstName": "Daniel",
  "lastName": "Ricciardo",
  "methodOfPayment": Object {
    "description": "Cash",
    "id": 910400001,
    "label": "Cash",
  },
  "mobilePhone": "07722 123456",
  "organisation": "Fishy Endeavours",
  "paymentSource": "Post Office Sales",
  "permitId": "test-permit-id",
  "postcode": "BS9 1HJ",
  "preferredMethodOfConfirmation": Object {
    "description": "Email",
    "id": 910400000,
    "label": "Email",
  },
  "preferredMethodOfNewsletter": Object {
    "description": "Prefer not to be contacted",
    "id": 910400003,
    "label": "Prefer not to be contacted",
  },
  "preferredMethodOfReminder": Object {
    "description": "Text",
    "id": 910400002,
    "label": "Text",
  },
  "premises": "14 Howecroft Court",
  "serialNumber": "14345-48457J",
  "startDate": "2021-06-15",
  "status": Object {
    "description": "Needs Review",
    "id": 910400000,
    "label": "Needs Review",
  },
  "street": "Eastmead Lane",
  "town": "Bristol",
  "transactionDate": undefined,
}
`;
